src/vm/i386/excepx86.cpp

```diff
@@ -1868,6 +1868,24 @@ NOINLINE LPVOID COMPlusEndCatchWorker(Thread * pThread)
     // Sync managed exception state, for the managed thread, based upon any active exception tracker
     pThread->SyncManagedExceptionState(fIsDebuggerHelperThread);
 
+    if (InlinedCallFrame::FrameHasActiveCall(pThread->m_pFrame))
+    {
+        // When unwinding an exception in ReadyToRun, the JIT_PInvokeEnd helper which unlinks the ICF from 
+        // the thread will be skipped. This is because unlike jitted code, each pinvoke is wrapped by calls
+        // to the JIT_PInvokeBegin and JIT_PInvokeEnd helpers, which push and pop the ICF on the thread. The
+        // ICF is not linked at the method prolog and unlined at the epilog when running R2R code. Since the
+        // JIT_PInvokeEnd helper will be skipped, we need to unlink the ICF here. If the executing method
+        // has another pinovoke, it will re-link the ICF again when the JIT_PInvokeBegin helper is called
+
+        MethodDesc* pPInvokeTargetMD = pThread->m_pFrame->GetFunction();
+        if (pPInvokeTargetMD != NULL && 
+            pPInvokeTargetMD->GetModule()->IsReadyToRun() &&
+            pPInvokeTargetMD->GetModule()->GetReadyToRunInfo()->IsReadyToRunEntryPoint(((InlinedCallFrame*)pThread->m_pFrame)->m_pCallerReturnAddress))
```

Comment:
This is not compatible with large version bubbles. With larger version bubbles, the MethodDesc may be from a different module than the module containing the code.

It may be better to check whether the code is managed by ReadyToRunJitManager

Reply:
Ok, that's exactly what I was looking for. This actually simplifies a whole lot!